runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,129 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_4x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_8x8x4_arm_64_dotprod)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x4x8_to_8x4x8_arm_64_dotprod(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 8 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 8));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int8x8_t vzero = vmov_n_s8(0);
+
+  int32x4_t acc[8];
+  for (int i = 0; i < 8; i++) {
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t rhs = vld1q_s8(rhs_ptr);
+    rhs_ptr += 16;
+
+    // We unpack int4s into individual int8s. To preserve signedness,
+    // int4s are moved to the upper 4-bits of each byte. This has the effect of
+    // multiplying each int4 by 2^4 = 16. To compensate, we divide the
+    // accumulator values by 16 before storing to memory.
+    // This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+    // kernels in https://github.com/google/XNNPACK.
+    int8x16_t rhs_0 = vshlq_n_s8(rhs, 4);
+    int8x16_t rhs_1 = vandq_s8(rhs, vmask);
+
+    if (M0 == 8) {
+      // 8x2 * 2x4 -> 8x4.
+      int8x16x2_t lhs_uzp[2];
+      for (int i = 0; i < 2; ++i) {
+        // We load the LHS de-interleaved to match the layout of `rhs_0` and
+        // `rhs_1`, where `rhs_0` contains odd-numbered channels and `rhs_1`
+        // contains even-numbered channels.
+        lhs_uzp[i] = vld2q_s8(lhs_ptr);
```

Comment:
unzip on lhs is a cute trick, but this could be removed if you use planar packed 4 bit and only use rhs_0 for the first tile and rhs_1 for the next tile.